### PR TITLE
TGP-2035 - refactor categorisation

### PR DESCRIPTION
**Pull request by ChrisCookOC**

### CODE REVIEW

File: conf/application.conf
Comment (besscerule) [inline conf/application.conf]: should this be included?

File: conf/app.routes
Comment (besscerule) [inline conf/app.routes]: these paths should have hyphens like

`/update-record/:recordId/categorisation/categorisation-preparation`

File: app/viewmodels/checkAnswers/AssessmentsSummary.scala
Comment (besscerule) [inline app/viewmodels/checkAnswers/AssessmentsSummary.scala]: given this is a boolean flag maybe we could change is to isReassementAnswer - or something that makes more sense?

File: app/viewmodels/checkAnswers/AssessmentsSummary.scala
Comment (besscerule) [inline app/viewmodels/checkAnswers/AssessmentsSummary.scala]: this feels a bit non-functional with two if statements assessing the same thing, pretty sure you could shoehorn most of this into line 50 so you would only need to check once

File: app/viewmodels/checkAnswers/CategorySummary.scala
Comment (besscerule) [inline app/viewmodels/checkAnswers/CategorySummary.scala]: you use this url twice so have it in a variable

File: app/viewmodels/checkAnswers/HasSupplementaryUnitSummary.scala
Comment (besscerule) [inline app/viewmodels/checkAnswers/HasSupplementaryUnitSummary.scala]: remove this

File: app/viewmodels/checkAnswers/LongerCommodityCodeSummary.scala
Comment (besscerule) [inline app/viewmodels/checkAnswers/LongerCommodityCodeSummary.scala:29]: this is way nicer! :)

File: app/viewmodels/checkAnswers/SupplementaryUnitSummary.scala
Comment (besscerule) [inline app/viewmodels/checkAnswers/SupplementaryUnitSummary.scala]: remove this line

File: app/services/CategorisationService.scala
Comment (besscerule) [inline app/services/CategorisationService.scala]: you don't need to have this as a separate variable just flat map straight from it

Comment (besscerule) [reply app/services/CategorisationService.scala]: also - does this even need to be in a service? it feels quite a small thing it is doing and could be directly in the controller maybe?

File: app/services/CategorisationService.scala
Comment (besscerule) [inline app/services/CategorisationService.scala]: shouldn't this redirect to a JR page?

File: app/services/CategorisationService.scala
Comment (besscerule) [inline app/services/CategorisationService.scala:64]: i'm just assuming that you got this right here - i don't think you can simply the ifs anymore

File: app/services/CategorisationService.scala
Comment (besscerule) [inline app/services/CategorisationService.scala]: do you need to have these as separate variables?

File: app/services/CategorisationService.scala
Comment (besscerule) [inline app/services/CategorisationService.scala:117]: this made me giggle

**Review by besscerule — COMMENTED** [review]
some comments

File: app/viewmodels/checkAnswers/CategorySummary.scala
Comment (yohan-oc) [reply app/viewmodels/checkAnswers/CategorySummary.scala]: Fixed :)

File: app/viewmodels/checkAnswers/HasSupplementaryUnitSummary.scala
Comment (yohan-oc) [reply app/viewmodels/checkAnswers/HasSupplementaryUnitSummary.scala]: Removed :)

File: app/services/CategorisationService.scala
Comment (Ewan-James-Donovan) [reply app/services/CategorisationService.scala]: I think since this service is used in multiple places, it's a big enough thing to not repeat in two places

File: app/services/CategorisationService.scala
Comment (ChrisCookOC) [reply app/services/CategorisationService.scala]: The return type doesn't allow that, so the place that calls it needs to handle that.

Could change the return type but that complicates logic more I think?

File: app/services/CategorisationService.scala
Comment (Ewan-James-Donovan) [reply app/services/CategorisationService.scala:64]: Could kind of 'flatten' it with pattern matching if you think it's worth it: (DON'T ASSUME THIS CODE IS CORRECT)
```
categorisationInfo match {
  case info if info.categoryAssessments.isEmpty =>
    StandardGoodsNoAssessmentsScenario

  case info 
    if info.categoryAssessmentsThatNeedAnswers.isEmpty && 
       info.categoryAssessments.exists(_.isCategory1) =>
    Category1NoExemptionsScenario

  case info if info.categoryAssessmentsThatNeedAnswers.isEmpty =>
    Category2Scenario

  case _ =>
    calculateBasedOnAnswers(categorisationInfo, userAnswers, recordId)
}
```

File: conf/application.conf
Comment (ChrisCookOC) [reply conf/application.conf]: Definitely not 🙈 good spot :)

File: app/services/CategorisationService.scala
Comment (Ewan-James-Donovan) [reply app/services/CategorisationService.scala:64]: I'm not sure that's any nicer now I look at it actually :')

**Comment by Ewan-James-Donovan** [comment]
I'm very happy with this. The simplifications brought in and the separation of concerns is making my dev heart smile.

File: app/services/CategorisationService.scala
Comment (ChrisCookOC) [reply app/services/CategorisationService.scala:64]: That could work. Would hold off for now just because its one of two places that need to change for Niphls/Nirms which will mean it changes soon anyway :)

File: app/controllers/CategorisationResultController.scala
Comment (liandra-giacomine) [inline app/controllers/CategorisationResultController.scala]: why do we need it if it's unused?

File: app/controllers/SupplementaryUnitController.scala
Comment (liandra-giacomine) [inline app/controllers/SupplementaryUnitController.scala]: when this is "allowed" to be empty the option should be None, why are we doing getOrElse("") here? :)

Comment (ChrisCookOC) [reply app/controllers/SupplementaryUnitController.scala]: It seems to be driven by the view. Tempted to change it. I think if the view should decide what it displays rather than the controller having to care :)

File: app/controllers/CategorisationResultController.scala
Comment (ChrisCookOC) [reply app/controllers/CategorisationResultController.scala]: So yeah on this one that is right we can remove it

On some of the others, one of the functions is using it so its showing as unused in the class for some reason, so those ones need to be flagged unused to keep the compiler happy :)

File: conf/app.routes
Comment (ChrisCookOC) [reply conf/app.routes]: Fixed :)

File: app/viewmodels/checkAnswers/AssessmentsSummary.scala
Comment (ChrisCookOC) [reply app/viewmodels/checkAnswers/AssessmentsSummary.scala]: Fixed :)

File: app/viewmodels/checkAnswers/AssessmentsSummary.scala
Comment (ChrisCookOC) [reply app/viewmodels/checkAnswers/AssessmentsSummary.scala]: Combined the two statements for now, how does that look?

File: app/viewmodels/checkAnswers/SupplementaryUnitSummary.scala
Comment (ChrisCookOC) [reply app/viewmodels/checkAnswers/SupplementaryUnitSummary.scala]: Done :) , another casualty in Beth's war on the empty line 😆

File: app/services/CategorisationService.scala
Comment (ChrisCookOC) [reply app/services/CategorisationService.scala]: Cleaned up a bit :)

File: app/controllers/SupplementaryUnitController.scala
Comment (ChrisCookOC) [reply app/controllers/SupplementaryUnitController.scala]: Changed this now so the view deals with the option :) Also means its in one place rather than 4!

**Review by liandra-giacomine — CHANGES_REQUESTED** [review]
I think we can update a few of these if statements to be more clean

File: app/controllers/CategorisationResultController.scala
Comment (vinoocs) [inline app/controllers/CategorisationResultController.scala]: can be accessed without a profile and user can click on the links  please can we add profileAuth like in other controllers

File: app/controllers/CyaCategorisationController.scala
Comment (vinoocs) [inline app/controllers/CyaCategorisationController.scala]: Please can we add recordId for the log like we have in other places

File: app/controllers/HasSupplementaryUnitController.scala
Comment (vinoocs) [inline app/controllers/HasSupplementaryUnitController.scala]: we removed profileAuth?

Comment (ChrisCookOC) [reply app/controllers/HasSupplementaryUnitController.scala]: Oops good spot, added it back!

File: app/controllers/CyaCategorisationController.scala
Comment (ChrisCookOC) [reply app/controllers/CyaCategorisationController.scala]: Done :)

File: app/controllers/CategorisationResultController.scala
Comment (ChrisCookOC) [reply app/controllers/CategorisationResultController.scala]: Good shout, added :)

**Review by liandra-giacomine — DISMISSED** [review]

**Review by vinoocs — APPROVED** [review]

**Pull request merged**
